### PR TITLE
Adding Domain Model (broken into Definition and Runtime)

### PR DESCRIPTION
**Pull request by stjohnjohnson**
screen shot 2016-07-19 at 10 36 41

### CODE REVIEW

**Comment by nicolaifsf** [comment]
Maybe change the ordering of the stuff under matrix, with the code for doing the different version tests coming first, followed with the explanation of the number of different versions combinations will be done.  

Workflow something similar, the ordering might be a little easier to understand if the actual code was prior to the explanation. I was a little confused what you meant by deploy-east and deploy-west until I continued reading.

Under containers, what I was thinking of putting was something like:
"A Container runs a task in an isolated environment. This is done in order to determine dependency version compatibility, without affecting other tasks that may be running at the some time. This is implemented using docker containers."

With metadata, might be cool to show an example of how to pass metadata into sd.

The images are really helpful for understanding.

**Comment by stjohnjohnson** [comment]
@nicolaifsf updated

screen shot 2016-07-16 at 10 31 28 am

**Comment by nicolaifsf** [comment]
Looks good to me. Another question I was having was what other sections will exist. I see there's a home and architecture portion to the guide, but anything thats more specific, perhaps for the client usage and etc?

**Comment by stjohnjohnson** [comment]
@nicolaifsf absolutely, but I figured those would be different PRs.  I just created a bunch of issues to track things like that.

**Comment by nicolaifsf** [comment]
Gotcha gotcha. 👍

File: docs/architecture/domain.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+### Pipeline
+
+A pipeline represents a collection of [jobs] that share the same source code. These jobs are executed in the order defined by the [workflow]. Only one job is required to be defined in every pipeline:
+
+ - `main` - Job that runs on each change to the source code.  _Note: this is also the same job that runs on proposed changes via GitHub Pull Requests_
+
+### Workflow
+
+Workflow is the defined order that [jobs] will execute in. Jobs can be executed in parallel, series, and a combination of the two to allow for all possibilities. All jobs executed in a given workflow run share:
```
Comment (nkatzman) [inline docs/architecture/domain.md]: executed in parallel, series, `or` a combination of the two

File: docs/architecture/domain.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+ - Source code checked out to the same git commit from the `main` [build] that triggered this
+ - Access to the [metadata] from the `main` [build] that triggered this
```
Comment (nkatzman) [inline docs/architecture/domain.md]: Does that mean that they can modify the `metadata` from the `main` build? Or is just that they have RO access to this metadata?

If the second, maybe rephrase it as:
"A copy of metadata from the main build"

Also can a human trigger this job? If so, what metadata/git commit does it use? Is it the last successful `main` build?

File: docs/architecture/domain.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+There are two requirements for your pipeline's workflow:
+
+ - Must start with `main` (triggered when the git repository changes)
```
Comment (nkatzman) [inline docs/architecture/domain.md]: If it must start with `main` why require it to be explicitly set every time?

File: docs/architecture/domain.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+ - Must contain every job (no non-triggered jobs)
+
+In the following example, this is the flow that will happen:
```
Comment (nkatzman) [inline docs/architecture/domain.md]: In the following example of a `workflow` section, this is the flow:

I just get a little confused when I see yaml snippets. Like is this all you need in the yaml? or is this just a specific section?

File: docs/architecture/domain.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
+## Definition Model
+
+![Definition](assets/definition-model.png)
+
+### Pipeline
+
+A pipeline represents a collection of [jobs] that share the same source code. These jobs are executed in the order defined by the [workflow]. Only one job is required to be defined in every pipeline:
+
+ - `main` - Job that runs on each change to the source code.  _Note: this is also the same job that runs on proposed changes via GitHub Pull Requests_
```
Comment (tkyi) [inline docs/architecture/domain.md]: `- Default job which runs for each change made to the source code.`

File: docs/architecture/domain.md (hunk 1 above)
Comment (tkyi) [inline docs/architecture/domain.md]: +1

File: docs/architecture/domain.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+ - `deploy-west` triggers `validate-west`
+ - `deploy-east` triggers `validate-east`
+
+### Job
+
+A job consists of executing multiple sequential [steps] inside a specified [container]. If any step in the series fails, then the entire job is considered failed and subsequent steps will be skipped (unless configured otherwise).
+
+Jobs share four pieces of context between steps:
+
+ - Source Code - Checked out to a given revision at the start of the job
```
Comment (nkatzman) [inline docs/architecture/domain.md]: You use `commit` up above. Can you use that instead of `revision`  ?

**Comment by Filbird** [comment]
This might be a negligible suggestion, but I found the model easier to understand when definitions are listed in the order of increasing encompassment (e.g. Step->Job->Pipeline), similar to the documentation in internal Screwdriver. Then again, you already have these definitions hyperlinked, so a newcomer can easily navigate through this document.

Regardless, this looks good! I personally found the information to be helpful.

File: docs/architecture/domain.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+ - Filesystem - Any changes made by a step to the filesystem will be visible to subsequent steps
```
Comment (nkatzman) [inline docs/architecture/domain.md]: You already mention that Jobs share these things between steps, so no need to add the `will be visible to subsequent steps`

File: docs/architecture/domain.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+ - Container - Any processes started or stopped by a step will be visible to subsequent steps
+ - Metadata - Structured key/value storage of relevant information
+
+Jobs are started by changes in source code, as a trigger from an upstream job, or by human decision.
```
Comment (nkatzman) [inline docs/architecture/domain.md]: Does a trigger need to be from an upstream job, or can jobs trigger jobs cross pipelines?

File: docs/architecture/domain.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+## Runtime Model
+
+![Runtime](assets/runtime-model.png)
+
+### Build
+
+A build is a run of a given [job] at a specified git commit. These will always have at least one [task] associated with it. A build's status is the overall result of all of the related sub-tasks.
+
+### Task
```
Comment (nkatzman) [inline docs/architecture/domain.md]: I would put this above `Container` because container references it.

File: docs/architecture/domain.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+ - `NODE_VERSION=6` and `EXPRESS_VERSION=2`
+ - `NODE_VERSION=6` and `EXPRESS_VERSION=3`
+
+### Step
+
+A step is a named action that needs to be performed, usually a single shell command. If the command finishes with a non-zero exit code, the step is considered a failure.
+
+### Container
+
+A Container runs a [task] in an isolated environment. This is done in order to determine dependency version compatibility, without affecting other tasks that may be running at the some time. This is implemented using Docker containers.
```
Comment (nkatzman) [inline docs/architecture/domain.md]: what is dependency version compatibility?

File: docs/architecture/domain.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+## Runtime Model
+
+![Runtime](assets/runtime-model.png)
+
+### Build
+
+A build is a run of a given [job] at a specified git commit. These will always have at least one [task] associated with it. A build's status is the overall result of all of the related sub-tasks.
```
Comment (nkatzman) [inline docs/architecture/domain.md]: change `sub-tasks` to `tasks` because sub tasks makes it seem like there are tasks and subtasks

File: docs/architecture/domain.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+### Task
+
+Tasks are the [matrix] permutations of a [build] that execute in a [container].
```
Comment (nkatzman) [inline docs/architecture/domain.md]: `Tasks are the [matrix] combinations of a [job]`

File: docs/architecture/domain.md (hunk 9 above)
Comment (nkatzman) [inline docs/architecture/domain.md]: But now I see job references container. ahhhh circular dependenciesssss

File: docs/architecture/domain.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+A pipeline represents a collection of [jobs] that share the same source code. These jobs are executed in the order defined by the [workflow]. Only one job is required to be defined in every pipeline:
+
+ - `main` - Job that runs on each change to the source code.  _Note: this is also the same job that runs on proposed changes via GitHub Pull Requests_
+
+### Workflow
+
+Workflow is the defined order that [jobs] will execute in. Jobs can be executed in parallel, series, and a combination of the two to allow for all possibilities. All jobs executed in a given workflow run share:
+
+ - Source code checked out to the same git commit from the `main` [build] that triggered this
```
Comment (tkyi) [inline docs/architecture/domain.md]: - Source code checked out `from` the same git commit `as` the main [build] that triggered this ___

File: docs/architecture/domain.md (hunk 3 above)
Comment (tkyi) [inline docs/architecture/domain.md]: - Must start with main `job` (triggered when the git repository changes)

File: docs/architecture/domain.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+ - Source Code - Checked out to a given revision at the start of the job
+ - Filesystem - Any changes made by a step to the filesystem will be visible to subsequent steps
+ - Container - Any processes started or stopped by a step will be visible to subsequent steps
+ - Metadata - Structured key/value storage of relevant information
+
+Jobs are started by changes in source code, as a trigger from an upstream job, or by human decision.
+
+### Matrix
+
+A build matrix is a set of environment permutations for a given [job]. Permutations are usually used for testing against multiple [containers] or test types.
```
Comment (tkyi) [inline docs/architecture/domain.md]: [container]s

**Comment by tkyi** [comment]
What's the difference between a `Workflow` and a `Pipeline`?

File: docs/architecture/domain.md (hunk 10 above)
Comment (nicolaifsf) [inline docs/architecture/domain.md]: Regarding multiple versions of multiple dependencies being compatible with one another.  
This could be worded better. Thanks!

**Comment by nicolaifsf** [comment]
I believed the key distinction is between one being the "collection", and the other the defined ordering.  Correct me if I am mistaken please!

File: docs/architecture/domain.md (hunk 2 above)
Comment (stjohnjohnson) [inline docs/architecture/domain.md]: Updated.  But in terms of the human trigger, it would be just like a normal commit if starting the `main` job.  But for other jobs, the users would have to pick a previous `main` build to base it on.  How can we best describe that.

File: docs/architecture/domain.md (hunk 3 above)
Comment (stjohnjohnson) [inline docs/architecture/domain.md]: Good point, @nkatzman.  We can probably just assume it is after main.

File: docs/architecture/domain.md (hunk 2 above)
Comment (nkatzman) [inline docs/architecture/domain.md]: Very carefully.

Maybe here we just wanna say:
" Source code checked out to the same git commit"
"Access to metadata from a `main` build that triggered or was selected for this job's build"

**Comment by tkyi** [comment]
So is it like in Wercker where you can chain `pipelines` together in a `workflow` or is a Screwdriver `workflow` the stuff in a `screwdriver.yaml` which has a one to one mapping with a `pipeline`?
http://devcenter.wercker.com/learn/workflows/introduction.html

File: docs/architecture/domain.md (hunk 12 above)
Comment (stjohnjohnson) [inline docs/architecture/domain.md]: Ugh, this is the hard part.  Because you have two concepts - the definition and execution of the thing.
- Job - definition of steps to run with a matrix in containers
- Build - desired execution of a Job

Vs.
- Task - a permutation of a Job
- ???? - desired execution of a Task

File: docs/architecture/domain.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
+## Definition Model
+
+![Definition](assets/definition-model.png)
+
+### Pipeline
+
+A pipeline represents a collection of [jobs] that share the same source code. These jobs are executed in the order defined by the [workflow]. Only one job is required to be defined in every pipeline:
```
Comment (d2lam) [inline docs/architecture/domain.md]: "**At least one job** is required ...." sounds more clear

**Comment by stjohnjohnson** [comment]
`screwdriver.yaml` is a 1-1 mapping to a pipeline.
`pipeline` is a collection of jobs and the order in which to run them.
`workflow` is the desired order you want your jobs to run in (see above).

File: docs/architecture/domain.md [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+After the merge of a pull-request to master:
+
+ - `main` will run and trigger `publish`
+ - `publish` will trigger `deploy-west` and `deploy-east` in parallel
+ - `deploy-west` triggers `validate-west`
+ - `deploy-east` triggers `validate-east`
+
+### Pipeline
+
+A pipeline represents a collection of [job]s that share the same source code. These jobs are executed in the order defined by the [workflow]. The `main` job is required to be defined in every pipeline:
```
Comment (d2lam) [inline docs/architecture/domain.md]: `[job]s` looks kinda weird. Can we add `[jobs]` and just add that to point to the same link?

File: docs/architecture/domain.md (hunk 10 above)
Comment (nkatzman) [inline docs/architecture/domain.md]: I mean to say, Is that the only reason we are running in an isolated environment? Maybe it's better to say. "This is done in order to test the compatibility of code running in different environments with different versions"

And this also says "running at the some time" should be "same time"

File: docs/architecture/domain.md [hunk 17]
```diff
@@ -0,0 +1,117 @@
+## Definition Model
+
+![Definition](assets/definition-model.png)
```
Comment (d2lam) [inline docs/architecture/domain.md:3]: This graph is a little confusing to me. Why does one job split into multiple jobs?

File: docs/architecture/domain.md [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+ - [Container]
+ - [Metadata]
+
+Jobs are started by changes in [source code], as a trigger from the [workflow], or by human decision.
+
+### Matrix
+
+A build matrix is a set of environment permutations for a given [job]. Permutations are usually used for testing against multiple [container]s or test types.
+
+In this example, the job will run 4 times for a given [build]:
```
Comment (d2lam) [inline docs/architecture/domain.md]: Run in parallel?

File: docs/architecture/domain.md (hunk 16 above)
Comment (stjohnjohnson) [inline docs/architecture/domain.md]: Lol, I had that and @tkyi suggested I change it.

File: docs/architecture/domain.md (hunk 17 above)
Comment (stjohnjohnson) [inline docs/architecture/domain.md:3]: Not the same job, it's more jobs in the workflow.

File: docs/architecture/domain.md (hunk 16 above)
Comment (stjohnjohnson) [inline docs/architecture/domain.md]: Fixed.

File: docs/architecture/domain.md (hunk 17 above)
Comment (d2lam) [inline docs/architecture/domain.md:3]: Yeah but from the graph it doesn't look clear what "workflow" is. Maybe have a rectangle around all those 4 jobs and caption it as "workflow"?

Comment (stjohnjohnson) [inline docs/architecture/domain.md:3]: What would be a better way to draw that?  Should I put job names?

**Comment by stjohnjohnson** [comment]
@nkatzman what should we do about `tasks` because we have two concepts - the definition and execution of the thing.

Job - definition of steps to run with a matrix in containers
Build - desired execution of a Job

Vs.

Task - a permutation of a Job
???? - desired execution of a Task

**Comment by nkatzman** [comment]
@stjohnjohnson should we call it a "Run" ?

File: docs/architecture/domain.md (hunk 17 above)
Comment (tkyi) [inline docs/architecture/domain.md:3]: I agree with @d2lam, the graph is a little confusing. 
You could put numbers to indicate order of execution and/or put job names that indicate order (Job1, Job2, etc.)

File: docs/architecture/domain.md [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+
+A Container runs [steps] in an isolated environment. This is done in order to test the compatibility of code running in different environments with different versions, without affecting other [tasks] that may be running at the same time. This is implemented using Docker containers.
+
+### Job
+
+A job consists of executing multiple sequential [steps] inside a specified [container]. If any step in the series fails, then the entire job is considered failed and subsequent steps will be skipped (unless configured otherwise).
+
+Jobs work by checking out the [source code] to a specified commit, setting the desired environment variables, and executing the specified [steps].
+
+During the job, the executing [steps] share three pieces of context:
```
Comment (jer) [inline docs/architecture/domain.md:24]: Don't they also share environment?

**Comment by tkyi** [comment]
`docs/architecture/domain.md:54: R2 Line has trailing whitespace`

**Comment by tkyi** [comment]
Is it `Build #1.1`, `Build #1.2`, `Build #1.3` or `Tasks`?

Otherwise, LGTM 👍

File: docs/architecture/domain.md (hunk 19 above)
Comment (stjohnjohnson) [inline docs/architecture/domain.md:24]: Eh, we didn't implement that yet.  It has some crazy side-effects (as we've seen).

**Comment by stjohnjohnson** [comment]
Removed tasks, replaced with just builds.

File: docs/architecture/domain.md [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+ - `NODE_VERSION=4` and `TEST_TYPE=unit`
+ - `NODE_VERSION=4` and `TEST_TYPE=functional`
+ - `NODE_VERSION=6` and `TEST_TYPE=unit`
+ - `NODE_VERSION=6` and `TEST_TYPE=functional`
+
+### Workflow
+
+Workflow is the defined order that [jobs] will execute in after a successful [build] of the `main` job on the default branch. Jobs can be executed in parallel, series, or a combination of the two to allow for all possibilities. Workflow must contain all defined jobs in the pipeline.
+
+All jobs executed in a given workflow run share:
```
Comment (jer) [inline docs/architecture/domain.md]: All jobs executed in a given workflow share:

File: docs/architecture/domain.md [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+
+A pipeline represents a collection of [jobs] that share the same [source code]. These jobs are executed in the order defined by the [workflow].
+
+The `main` job is required to be defined in every pipeline as it is the one that builds for each change made to the source code (and proposed changes).
+
+## Runtime Model
+
+![Runtime](assets/runtime-model.png)
+
+### Build
```
Comment (jer) [inline docs/architecture/domain.md]: Can we move this up by job and rephrase it completely?

**Comment by jer** [comment]
👍  
timelyflowerydesertpupfish-size_restricted

File: docs/architecture/domain.md [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+
+Jobs are started by changes in [source code], as a trigger from the [workflow], or by human decision.
+
+#### Parallelization
+
+It is possible to parallelize a job by defining a matrix of environment variables. These are usually used for testing against multiple [containers] or test types.
+
+In this example job definition, 4 [builds] will run in parallel:
+```yaml
+image: node:((NODE_VERSION))
```
Comment (d2lam) [inline docs/architecture/domain.md]: `${NODE_VERSION}` ?

File: docs/architecture/domain.md [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+ - `NODE_VERSION=6` and `TEST_TYPE=functional`
+
+### Build
+
+A build is an instance of a running [job]. All builds are assigned a unique build number. With a basic job configuration, only one build of a job will be running at any given time. If a [job matrix] is configured, then there can be multiple builds running in parallel.
+
+A build can be in one of five different states:
+
+ - `queued` - Build is waiting for available resources
+ - `running` - Build is actively running on an executor
```
Comment (d2lam) [inline docs/architecture/domain.md:62]: Should these match with the status when we do a get? https://github.com/screwdriver-cd/data-schema/blob/master/model/build.js#L51-L55

running -> in progress
successful -> success 
...

File: docs/architecture/domain.md [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+        - series:
+            - deploy-east
+            - validate-east
+```
+
+After the merge of a pull-request to master:
+
+ - `main` will run and trigger `publish`
+ - `publish` will trigger `deploy-west` and `deploy-east` in parallel
+ - `deploy-west` triggers `validate-west`
```
Comment (d2lam) [inline docs/architecture/domain.md]: `will trigger` to be consistent

File: docs/architecture/domain.md [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+ - `deploy-east` triggers `validate-east`
```
Comment (d2lam) [inline docs/architecture/domain.md]: `will trigger` to be consistent

File: docs/architecture/domain.md (hunk 23 above)
Comment (stjohnjohnson) [inline docs/architecture/domain.md:62]: Yeah, we'll be updating them.

File: docs/architecture/domain.md [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+
+Jobs work by checking out the [source code] to a specified commit, setting the desired environment variables, and executing the specified [steps].
+
+During the job, the executing [steps] share three pieces of context:
+
+ - Filesystem
+ - [Container]
+ - [Metadata]
+
+Jobs are started by changes in [source code], as a trigger from the [workflow], or by human decision.
```
Comment (tkyi) [inline docs/architecture/domain.md]: Jobs can be started automatically by changes made in the source code or triggered through the workflow. Jobs can also be started manually through the UI.

File: docs/architecture/domain.md (hunk 22 above)
Comment (tkyi) [inline docs/architecture/domain.md]: or `{{NODE_VERSION}}`?

**Comment by tkyi** [comment]
👍  W┴פ˥
giphy
